components/constellation/network_listener.rs

```diff
@@ -0,0 +1,115 @@
+            req_init.url_list.push(req_init.url.clone());
+        }
+
+        let mut listener = NetworkListener {
+            res_init: self.res_init.clone(),
+            req_init: req_init.clone(),
+            resource_threads: self.resource_threads.clone(),
+            sender: self.sender.clone(),
+            pipeline_id: self.pipeline_id.clone(),
+            should_send: false,
+        };
+
+        ROUTER.add_route(ipc_receiver.to_opaque(), box move |message| {
+            let msg = message.to();
+            match msg {
+                Ok(FetchResponseMsg::ProcessRequestReferrer(referrer, referrer_policy)) => {
+                    listener.req_init.referrer_url = match referrer.clone() {
+                        Referrer::ReferrerUrl(servo_url) => Some(servo_url),
+                        _ => None,
+                    };
+                    listener.req_init.referrer = Some(referrer);
+                    listener.req_init.referrer_policy = referrer_policy;
+                }
+                Ok(FetchResponseMsg::ProcessResponse(res)) => listener.check_redirect(res),
+                Ok(msg_) => listener.send(msg_),
+                _ => {}
+            };
+        });
+
+        let msg = match self.res_init {
+            Some(ref res_init_) => CoreResourceMsg::FetchRedirect(
+                                   req_init.clone(),
+                                   res_init_.clone(),
+                                   ipc_sender),
+            None => CoreResourceMsg::Fetch(
+                    req_init,
+                    ipc_sender)
+        };
+
+        if let Err(e) = self.resource_threads.sender().send(msg) {
+            warn!("Exit resource thread ({})", e);
+        }
+    }
+
+    fn check_redirect(&mut self,
+                      message: Result<(FetchMetadata), NetworkError>) {
+        match message {
+            Ok(res_metadata) => {
+                let metadata = match res_metadata {
+                    FetchMetadata::Filtered { ref unsafe_, .. } => unsafe_,
+                    FetchMetadata::Unfiltered(ref m) => m,
+                };
+
+                match metadata.headers {
+                    Some(ref headers) if headers.has::<Location>() => {
+                        if self.res_init.is_some() {
+                            self.req_init.url_list.push(metadata.final_url.clone());
+                        }
+
+                        self.res_init = Some(ResponseInit {
+                            url: metadata.final_url.clone(),
+                            headers: headers.clone().into_inner(),
+                            referrer: metadata.referrer.clone(),
+                        });
+
+                        self.initiate_fetch();
+                    },
+                    _ => {
+                        // when response does not redirect
+                        self.should_send = true;
+                        self.send(FetchResponseMsg::ProcessResponse(Ok(res_metadata.clone())));
```

Comment:
I don't think this willl prevent the clone, as `res_metadata` has already been borrowed in the match block to exract `metadata` from it, a few lines above this code.